PR for Feature Request: Service Fabric configuration source #9

chuanboz opened the pull request:
I've added the support as well as added the unit test to cover the new configuration source support, please help to take a look.

chuanboz committed:
Feature Request: Service Fabric configuration source #9

Part 1 check-in, add the support. still pending adding more tests.

chuanboz committed:
basic test added.

chuanboz committed:
validated config actions as well as other cases

chuanboz committed:
update dependency

amanbha commented on src/netstandard/Microsoft.ServiceFabric.AspNetCore.Configuration/ServiceFabricConfigurationExtensions.cs:
```diff
@@ -0,0 +1,61 @@
+// Licensed under the MIT License (MIT). See License.txt in the repo root for license information.
+// ------------------------------------------------------------
+
+namespace Microsoft.ServiceFabric.AspNetCore.Configuration
```
>namespace [](start = 0, length = 9)

This file is not needed in netstandard folder, netstandard .csproj will use the file from full framework folder.

chuanboz replied:
fixed.

amanbha commented on src/Microsoft.ServiceFabric.AspNetCore.Configuration/ServiceFabricConfigurationProvider.cs:
```diff
@@ -0,0 +1,80 @@
+                    // With this reason, will treat encrypted value the same as plain text,
+                    // user will need to handle encrypted string separately to compliant with security best practice.
+                    // Data[$"{section.Name}{ConfigurationPath.KeyDelimiter}{param.Name}"] = param.IsEncrypted ? param.DecryptValue().ToUnsecureString() : param.Value;
+                    data[$"{section.Name}{ConfigurationPath.KeyDelimiter}{param.Name}"] = param.Value;
```
Since multiple config packages are supported, add package name here as well, else a setting with same section name and param name will override value from other package. I will sync up with you reg. this in a skype call (if possible).

chuanboz replied:
Adding package name to the key will make the typical usage much harder, as you will need to add the pacage name to every time when you GetSection.

Is it OK to introduce an Options switch and use the switch to turn this behavior On and Off, and default to make it off?

amanbha replied:
Sure, but lets add package name as default and use the switch to not add it.

chuanboz replied:
I've added the default to true if there is multiple package, and will be false if there is only single package.

amanbha commented on src/Microsoft.ServiceFabric.AspNetCore.Configuration/ServiceFabricConfigurationProvider.cs:
```diff
@@ -0,0 +1,80 @@
+        {
+            if (reload)
+            {
+                this.Data.Clear();  // Rememove the old keys on re-load
```
>Rememove [](start = 39, length = 8)

typo #Closed

chuanboz replied:
thanks, fixed. #Closed

chuanboz committed:
Added options and addressed feedback from Aman

Added ServiceFabricConfigurationOptions, and moved all switch and configuraton action to that class, and changed default to include package name if that's multiple packages.

chuanboz commented:
@amanbha , I've addressed your feedback and have updated a new version which introduced options for advanced scenarios.

RamjotSingh commented:
@chuanboz Based on if there is one config objects or more the key will look something like configName:sectionName:paramName vs sectionName:paramName. For simplicity sake would it make sense to just always have configName:sectionName:paramName? This way if someone adds a new config they won't have to go back and modify their existing code.

amanbha commented on src/Microsoft.ServiceFabric.AspNetCore.Configuration/ServiceFabricConfigurationOptions.cs:
```diff
@@ -0,0 +1,129 @@
+        /// The default is true, and will include the package name, section name and param namein configuraton package.
+        /// You could set this to true for multiple config packages to avoid a setting with same section name and param name being override from other package.
+        /// </summary>
+        public bool IncludePackageName { get; set; }
```
>IncludePackageName [](start = 20, length = 18)

Thinking more about, We can just remove this flag and always add config package name to the key, if a separate behavior is needed, user can always provide the ExtractKeyFunc. This would keep the default usage simple and is consistant when user will have single or multiple packages. Users dont need to worry about it wether they have one or more packages.

chuanboz replied:
thanks @RamjotSingh  and @amanbha , feedback received so I remove the condition logic on the default and always make default to true.

I'm stuggle on whether to completely remove the property of IncludePackageName, I think it still make sense to have IncludePackageName, given it is default to be true, it won't causing confusing. If we need to remove the IncludePackageName I would then also consider remove DecryptValue flag as the same reason could be given that user could provide ExtractValueFunc.

amanbha replied:
I understand your thinking about keeping it, If you want to keep the property , its fine, but I would suggest to remove the statement in ServiceFabricConfigurationExtensions to set it to false when there is only one config package. This way user will nto have to change their code when they add more config packages.

---
In reply to: [219564022](https://github.com/Microsoft/service-fabric-aspnetcore/pull/58#discussion_r219564022) [](ancestors = 219564022)

chuanboz replied:
Sure, I already removed that in extensions and fixed the UT, will send new PR soon.

chuanboz replied:
new PR send that i addressed the feedback. let me know if there are further feedback and I will address them soon also.

chuanboz committed:
Changed default value of IncludePackageName to true

addressed feedback to change default value of IncludePackageName to true, fixed UT because of this change, and also updated the comment in the options class.

RamjotSingh commented:
@chuanboz Not exactly a feedback on the PR but did you try using the deserialization of section directly? Json settings (appSettings.json) allows you to just give a strongly typed class and use a section to fill it up I am wondering if that same thing will work here.

chuanboz commented:
> @chuanboz Not exactly a feedback on the PR but did you try using the deserialization of section directly? Json settings (appSettings.json) allows you to just give a strongly typed class and use a section to fill it up I am wondering if that same thing will work here.

I have not tried that, but my team's plan is to just put the json file directly in configuration package, rather than embeded the json file in xml file. But if do so I could see the ConfigAction could be used to serve that purpose to extract the config and populate to the IConfiguration.

chuanboz commented:
@RamjotSingh , I think I might misunderstand you just now, for deserialization of section directly we do use that, and that will be available via any IConfiguration. that's exactly the purpose of this provider, to enable you. I will update the PR to include one such usage in test soon.

RamjotSingh commented:
> @RamjotSingh , I think I might misunderstand you just now, for deserialization of section directly we do use that, and that will be available via any IConfiguration. that's exactly the purpose of this provider, to enable you. I will update the PR to include one such usage in test soon.

Thanks :)

chuanboz committed:
Add test case to demo to bind to object directly

chuanboz commented:
added the usage in TestHappyCase.

```C#
           // basic validate to bind to a class directly
            // Note, in asp.net core 2.1 you could use the more simple ConfigurationBinder.Get<T> binds and returns the specified type instance directly.
            // Get<T> is more convenient than using Bind but will require .net core version higher than 1.0
            var person = new Person();
            config.GetSection("Config:Section1").Bind(person);

            person.Name.Should().Be("Xiaoxiao");
            person.Age.Should().Be(6);
            person.Gender.Should().Be("M");
```

chuanboz commented:
@amanbha , let me know if you have further comment for this PR.

amanbha commented:
@chuanboz , it looks good, thanks for adding this. I will merge it into develop branch, For release logistics, I will sync up with @vturecek to see which release we want to put this in as out current release is locked down.

RamjotSingh commented:
@amanbha  Can you merge this into develop so people can take dependency on this by submoduling? Also would it be possible to release it to myget or some other daily release feeds?

amanbha commented:
@RamjotSingh I will merge it to develop, we don't have a daily release feed but you can consume the binary directly in your code if needed or push it to your own custom feeds.

RamjotSingh commented:
@amanbha I guess I will drop the nuget in my own repo directly for now. Thanks.
